src/client/datascience/jupyter/kernels/kernelSelector.ts

```diff
@@ -259,8 +269,11 @@ export class KernelSelector implements IKernelSelectionUsage {
         ]);
 
         // First check for a live active session.
-        if (notebookMetadata && notebookMetadata.id) {
-            const session = sessions?.find((s) => s.kernel.id === notebookMetadata?.id);
+        const preferredKernelId = resource
+            ? this.preferredRemoteKernelIdProvider.getPreferredRemoteKernelId(resource)
```

Comment:
Use separate class to find the remote session (the kernelId is tracked against the Uri in that class, thi way we don't need to pass & store id as part of metadata, removing)